client/core/trade.go

```diff
@@ -1405,60 +1405,95 @@ func (c *Core) tick(t *trackedTrade) (assetMap, error) {
 	t.mtx.Lock()
 	defer t.mtx.Unlock()
 	var swaps, redeems, refunds []*matchTracker
-	assets := make(assetMap)
-	errs := newErrorSet(t.dc.acct.host + " tick: ")
+	assets := make(assetMap) // callers expect non-nil map even on error
 
 	// Check all matches for and resend pending requests as necessary.
 	c.resendPendingRequests(t)
 
 	// Make sure we have a redemption fee suggestion cached.
 	t.cacheRedemptionFeeSuggestion()
 
-	// Check all matches and send swap, redeem or refund as necessary.
 	var sent, quoteSent, received, quoteReceived uint64
-	for _, match := range t.matches {
+	checkMatch := func(match *matchTracker) error {
 		side := match.Side
 		if (side == order.Maker && match.Status >= order.MakerRedeemed) ||
 			(side == order.Taker && match.Status >= order.MatchComplete) {
-			continue
+			return nil
 		}
 		if match.Address == "" {
-			continue // a cancel order match
+			return nil // a cancel order match
 		}
 		if !t.matchIsActive(match) {
-			continue // either refunded or revoked requiring no action on this side of the match
+			return nil // either refunded or revoked requiring no action on this side of the match
 		}
-		switch {
-		case t.isSwappable(c.ctx, match):
+
+		// The trackedTrade mutex is locked, so we must not hang forever. Give
+		// this a generous timeout because it may be necessary to retrieve full
+		// blocks, and catch timeout/shutdown after each check.
+		ctx, cancel := context.WithTimeout(c.ctx, 30*time.Second)
+		defer cancel()
+
+		if t.isSwappable(ctx, match) {
 			c.log.Debugf("Swappable match %s for order %v (%v)", match, t.ID(), side)
 			swaps = append(swaps, match)
 			sent += match.Quantity
 			quoteSent += calc.BaseToQuote(match.Rate, match.Quantity)
+			return nil
+		}
+		if ctx.Err() != nil { // may be here because of timeout or shutdown
+			return ctx.Err()
+		}
 
-		case t.isRedeemable(c.ctx, match):
+		if t.isRedeemable(ctx, match) {
 			c.log.Debugf("Redeemable match %s for order %v (%v)", match, t.ID(), side)
 			redeems = append(redeems, match)
 			received += match.Quantity
 			quoteReceived += calc.BaseToQuote(match.Rate, match.Quantity)
+			return nil
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 
 		// Check refundability before checking if to start finding redemption.
 		// Ensures that redemption search is not started if locktime has expired.
 		// If we've already started redemption search for this match, the search
 		// will be aborted if/when auto-refund succeeds.
-		case t.isRefundable(match):
+		if t.isRefundable(ctx, match) {
 			c.log.Debugf("Refundable match %s for order %v (%v)", match, t.ID(), side)
 			refunds = append(refunds, match)
+			return nil
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 
-		case t.shouldBeginFindRedemption(c.ctx, match):
+		if t.shouldBeginFindRedemption(ctx, match) {
```

Comment:
I've done a lot more refactoring to `tick` in https://github.com/decred/dcrdex/pull/1739